### PR TITLE
[r] Updated groups read/write support via libtiledbsoma

### PR DESCRIPTION
**Pull request by eddelbuettel**
**Issue and/or context:**

This PR updates the read and write operations for group and their metadata to the new C++ underpinnings.

**Changes:**

New code. Existing tests.

**Notes for Reviewer:**

[SC 54851](https://app.shortcut.com/tiledb-inc/story/54851/r-c-support-group-read-write-via-libtiledbsoma)

This may take a round or two of reviews as some comments are still in the code.  Tests now pass so it is ready.  It has also been rebased, and there is still a local/interim test file in there I will remove once the first set of CI tests is done.

### CODE REVIEW

File: apis/r/R/TileDBGroup.R [hunk 1]
```diff
@@ -60,22 +65,25 @@ TileDBGroup <- R6::R6Class(
...
-        private$.tiledb_group <- tiledb::tiledb_group(
-          self$uri,
-          type = mode,
-          ctx = private$.tiledb_ctx
-        )
+        spdl::debug("[TileDBGroup$open] Opening {} '{}' in {} mode", self$class(), self$uri, mode)
+        #private$.tiledb_group <- tiledb::tiledb_group(self$uri, type = mode, ctx = private$.tiledb_ctx)
+        #print(str(private$.soma_context))
+        private$.tiledb_group <- c_group_open(self$uri, type = mode, ctx = soma_context())#private$.some_context)
+        #group_open()
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:72]: Can we clean out the commented code?
```suggestion
        spdl::debug("[TileDBGroup$open] Opening {} '{}' in {} mode", self$class(), self$uri, mode)
        private$.tiledb_group <- c_group_open(self$uri, type = mode, ctx = soma_context())#private$.some_context)
```

File: apis/r/R/RcppExports.R [hunk 2]
```diff
@@ -13,6 +13,46 @@ writeArrayFromArrow <- function(uri, naap, nasp, arraytype = "", config = NULL,
     invisible(.Call(`_tiledbsoma_writeArrayFromArrow`, uri, naap, nasp, arraytype, config, tsvec))
 }
 
+#' Create a group
+#' @param uri The array URI
+#' @param ctxxp An external pointer to the SOMAContext wrapper
+#' @export
+c_group_create <- function(uri, type, ctxxp, timestamp = NULL) {
+    invisible(.Call(`_tiledbsoma_c_group_create`, uri, type, ctxxp, timestamp))
+}
```
Comment (mojaveazure) [inline apis/r/R/RcppExports.R:22]: Do we need this exported?

File: apis/r/R/TileDBGroup.R [hunk 3]
```diff
@@ -32,10 +32,15 @@ TileDBGroup <- R6::R6Class(
...
+      spdl::debug("[TileDBGroup$create] Creating new {} at '{}' at {}",
+                  self$class(), self$uri, self$tiledb_timestamp)
 
-      self$open("WRITE", internal_use_only = "allowed_use")
+      #tiledb::tiledb_group_create(uri = self$uri, ctx = private$.tiledb_ctx)
+      private$.soma_context <- soma_context()  # FIXME via factory and paramater_config
+      c_group_create(self$uri, self$class(), private$.soma_context,
+                     self$.tiledb_timestamp_range)  ## FIXME: use to be added accessor
+
+      #self$open("WRITE", internal_use_only = "allowed_use") # FIXME
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:43]: Can we clean out the commented code?
```suggestion
      private$.soma_context <- soma_context()  # FIXME via factory and paramater_config
      c_group_create(self$uri, self$class(), private$.soma_context,
                     self$.tiledb_timestamp_range)  ## FIXME: use to be added accessor
```

File: apis/r/R/TileDBGroup.R [hunk 4]
```diff
@@ -60,22 +65,25 @@ TileDBGroup <- R6::R6Class(
...
-        cfg["sm.group.timestamp_end"] <- format(round(as.numeric(private$.group_open_timestamp) * 1000),
-                                                scientific = FALSE, big.mark = "")
-        private$.tiledb_group <- tiledb::tiledb_group(self$uri, type = mode,
-                                                      ctx = private$.tiledb_ctx, cfg = cfg)
+        #cfg <- tiledb::config(private$.tiledb_ctx)
+        #cfg["sm.group.timestamp_end"] <- format(round(as.numeric(private$.group_open_timestamp) * 1000),
+        #                                        scientific = FALSE, big.mark = "")
+        #private$.tiledb_group <- tiledb::tiledb_group(self$uri, type = mode,
+        #                                              ctx = private$.tiledb_ctx, cfg = cfg)
+        #print(class(private$.soma_context)) ## -- is externalpointer
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:84]: Can we clean out the commented code?
```suggestion
```

File: apis/r/R/TileDBGroup.R [hunk 5]
```diff
@@ -98,7 +106,8 @@ TileDBGroup <- R6::R6Class(
         }
 
         spdl::debug("Closing {} '{}'", self$class(), self$uri)
-        tiledb::tiledb_group_close(private$.tiledb_group)
+        ##tiledb::tiledb_group_close(private$.tiledb_group)
+        c_group_close(private$.tiledb_group)
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:110]: Can we clean out the commented code?
```suggestion
        c_group_close(private$.tiledb_group)
```

File: apis/r/R/TileDBGroup.R [hunk 6]
```diff
@@ -133,16 +142,22 @@ TileDBGroup <- R6::R6Class(
...
-        uri = uri,
-        relative = relative,
-        name = name
-      )
+      #tiledb::tiledb_group_add_member(
+      #  grp = private$.tiledb_group,
+      #  uri = uri,
+      #  relative = relative,
+      #  name = name
+      #)
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:156]: Can we clean out the commented code?
```suggestion
```

File: apis/r/R/TileDBGroup.R [hunk 7]
```diff
@@ -195,10 +212,8 @@ TileDBGroup <- R6::R6Class(
       stopifnot(is_scalar_character(name))
       private$check_open_for_write()
 
-      tiledb::tiledb_group_remove_member(
-        grp = private$.tiledb_group,
-        uri = name
-      )
+      #tiledb::tiledb_group_remove_member(grp = private$.tiledb_group, uri = name)
+      c_group_remove_member(private$.tiledb_group, name)
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:216]: Can we clean out the commented code?
```suggestion
      c_group_remove_member(private$.tiledb_group, name)
```

File: apis/r/R/TileDBGroup.R [hunk 8]
```diff
@@ -365,21 +388,27 @@ TileDBGroup <- R6::R6Class(
     # list with names: name, uri, and type.
     get_all_members_uncached_read = function(group_handle) {
 
-      count <- tiledb::tiledb_group_member_count(group_handle)
+      #count <- tiledb::tiledb_group_member_count(group_handle)
+      count <- c_group_member_count(group_handle)
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:392]: Can we clean out the commented code?
```suggestion
      count <- c_group_member_count(group_handle)
```

File: apis/r/R/TileDBGroup.R [hunk 9]
```diff
@@ -365,21 +388,27 @@ TileDBGroup <- R6::R6Class(
...
-          nm = c("type", "uri", "name")
-        )
-      }
+      #for (i in seq_len(count)) {
+      #  members[[i]] <- setNames(
+      #    # returns character vector with <type, uri, name>
+      #    object = as.list(tiledb::tiledb_group_member(group_handle, i - 1L)),
+      #    nm = c("type", "uri", "name")
+      #  )
+      #}
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:404]: Can we clean out the commented code?
```suggestion
```

File: apis/r/R/TileDBGroup.R [hunk 10]
```diff
@@ -365,21 +388,27 @@ TileDBGroup <- R6::R6Class(
...
 
       # Key the list by group member name
-      names(members) <- vapply_char(members, FUN = getElement, name = "name")
+      #names(members) <- vapply_char(members, FUN = getElement, name = "name")
+
+      ## names list of lists
+      members <- c_group_members(group_handle)
+
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:411]: Can we clean out the commented code?
```suggestion
      ## names list of lists
      members <- c_group_members(group_handle)
```

File: apis/r/R/TileDBGroup.R [hunk 11]
```diff
@@ -401,8 +430,15 @@ TileDBGroup <- R6::R6Class(
         # FIXME: when we add write timestamps we should open this temp handle with tiledb_timestamp
         # too. The stopifnot is currently "unreachable" since open() stops if called with WRITE
         # mode and non-null tiledb_timestamp.
-        stopifnot("FIXME" = is.null(private$.group_open_timestamp))
-        group_handle <- tiledb::tiledb_group(self$uri, type = "READ", ctx = private$.tiledb_ctx)
+        #stopifnot("FIXME" = is.null(private$.group_open_timestamp))
+        # group_handle <- tiledb::tiledb_group(self$uri, type = "READ", ctx = private$.tiledb_ctx)
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:434]: Can we clean out the commented code?
```suggestion
```

File: apis/r/R/TileDBGroup.R [hunk 12]
```diff
@@ -415,7 +451,8 @@ TileDBGroup <- R6::R6Class(
       }
 
       if (private$.mode == "WRITE") {
-        tiledb::tiledb_group_close(group_handle)
+        #tiledb::tiledb_group_close(group_handle)
+        c_group_close(group_handle)
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:455]: Can we clean out the commented code?
```suggestion
        c_group_close(group_handle)
```

File: apis/r/R/TileDBGroup.R [hunk 13]
```diff
@@ -484,13 +521,17 @@ TileDBGroup <- R6::R6Class(
       # we must open a temporary handle for read, to fill the cache.
       group_handle <- private$.tiledb_group
       if (private$.mode == "WRITE") {
-        group_handle <- tiledb::tiledb_group(self$uri, type = "READ", ctx = private$.tiledb_ctx)
-      }
+        # group_handle <- tiledb::tiledb_group(self$uri, type = "READ", ctx = private$.tiledb_ctx)
+        group_handle <- c_group_open(self$uri, type ="READ", ctx = private$.soma_context,
+                                     private$.tiledb_timestamp_range)
```
Comment (mojaveazure) [inline apis/r/R/TileDBGroup.R:526]: Can we clean out the commented code?
```suggestion
        group_handle <- c_group_open(self$uri, type ="READ", ctx = private$.soma_context,
                                     private$.tiledb_timestamp_range)
```

**Review by mojaveazure — APPROVED** [review]
LGTM; there's some places where we could clean up some now-unused commented code (marked as suggestions so hopefully you can just click-and-apply), and one function that's marked as export that I'm not sure needs to be exported

File: apis/r/R/RcppExports.R (hunk 2 above)
Comment (eddelbuettel) [reply apis/r/R/RcppExports.R:22]: Good catch!  That is copypasta.  None of the c_group_* should need to be user-facing.  I will clean that up.

Comment (eddelbuettel) [reply apis/r/R/RcppExports.R]: Done!  Let's see if CI still likes it.

**Pull request merged**